### PR TITLE
Theme customization

### PR DESCRIPTION
**Pull request by AbijithS-aot**
# Issue Tracking

JIRA: 
Issue Type: BUG/ FEATURE

# Changes

created apis for theme customization
# How has the change been tested?

# Screenshots (if applicable)

image
image
image

### CODE REVIEW

**Comment by AbijithS-aot** [comment]
will create another pr with test cases

File: forms-flow-api/migrations/versions/95387de7067e_theme.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+"""theme customization
```
Comment (sumesh-aot) [inline forms-flow-api/migrations/versions/95387de7067e_theme.py:1]: Why do we have 2 migrations for same table creation? Can you please fix this

File: forms-flow-api/migrations/versions/0e526ddca7b3_theme.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+revision = '0e526ddca7b3'
+down_revision = 'fdfe787a197c'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table('theme_customization',
+    sa.Column('id', sa.Integer(), nullable=False),
```
Comment (sumesh-aot) [inline forms-flow-api/migrations/versions/0e526ddca7b3_theme.py:22]: I think we can have the table name as themes

File: forms-flow-api/migrations/versions/0e526ddca7b3_theme.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    sa.Column('logo_name', sa.String(length=50), nullable=False),
+    sa.Column('logo_type', sa.String(length=100), nullable=False),
+    sa.Column('value', sa.String(length=20), nullable=False),
```
Comment (sumesh-aot) [inline forms-flow-api/migrations/versions/0e526ddca7b3_theme.py:25]: Is the value intended for logo ? Please rename it to make it more clear.
Also value can be base64 string, so length 20 wouldn't work. Make it flexible by not adding length.

File: forms-flow-api/src/formsflow_api/models/theme.py [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    application_title = db.Column(db.String(50), nullable=False)
+    theme = db.Column(JSON, nullable=False)
+    tenant = db.Column(db.String(20), nullable=True)
+
+    @classmethod
+    def create_theme(cls, theme_info: dict):
+        """Create new theme."""
+        if theme_info:
+            theme = cls()
+            theme.created_by = theme_info["created_by"]
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/theme.py:26]: Please use .get here to make it consistent and to rule out KeyError.

File: forms-flow-api/src/formsflow_api/models/theme.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            theme.logo_name = theme_info.get("logo_name")
+            theme.logo_type = theme_info.get("logo_type")
+            theme.value = theme_info.get("value")
+            theme.application_title = theme_info.get("application_title")
+            theme.tenant = theme_info.get("tenant")
+            theme.theme = theme_info.get("theme")
+            theme.save()
+            return theme
+        return None
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/theme.py:35]: We should assert the incoming payload and make sure None value is not passed to the method.

File: forms-flow-api/src/formsflow_api/models/theme.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+        )
+        self.commit()
+
+    @classmethod
+    def get_theme(cls, tenant: str = None):
+        """Find application that matches the provided id."""
+        # For multi tenant setup there would be multiple records in this table,
+        # so match with tenant and return the record.
+        # For a non-multi tenant setup there SHOULD be only one record in this table, so return the record
+        if tenant:
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/models/theme.py:57]: ideally we can refactor this and have only one return statement

File: forms-flow-api/src/formsflow_api/resources/__init__.py [hunk 7]
```diff
@@ -61,3 +62,4 @@
 API.add_namespace(KEYCLOAK_ROLES_API, path="/roles")
 API.add_namespace(FORM_EMBED_API, path="/embed")
 API.add_namespace(INTEGRATION_API, path="/integrations")
+API.add_namespace(THEME_CUSTOMIZATION_API, path="/theme")
```
Comment (sumesh-aot) [inline forms-flow-api/src/formsflow_api/resources/__init__.py:65]: Can we make it plural themes ?

File: forms-flow-api/src/formsflow_api/services/theme.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        theme_customization = Themes.create_theme(data)
+        result = theme_schema.dump(theme_customization)
+        return result
+
+    @staticmethod
+    def get_theme(tenant_key):
+        """Return theme using tenant key else default theme."""
+        theme = Themes.get_theme(tenant_key)
+        theme_schema = ThemeCustomizationSchema()
+        result = theme_schema.dump(theme)
```
Comment (auslin-aot) [inline forms-flow-api/src/formsflow_api/services/theme.py:29]: Initialize a common ThemeCustomizationSchema() object and utilize the object within functions

File: forms-flow-api/src/formsflow_api/resources/theme.py [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+        "applicationTitle": fields.String(),
+        "theme": fields.Raw(),
+    },
+)
+
+
+@cors_preflight("GET, POST,PUT, OPTIONS")
+@API.route("", methods=["GET", "POST", "PUT", "OPTIONS"])
+class ThemeCustomizationResource(Resource):
+    """Resource to create theme."""
```
Comment (auslin-aot) [inline forms-flow-api/src/formsflow_api/resources/theme.py:37]: Please update the docstring to provide clear descriptions.

**Review by sumesh-aot — APPROVED** [review]

File: forms-flow-api/src/formsflow_api/services/theme.py [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+        theme = Themes.get_theme(tenant_key)
+        result = theme_schema.dump(theme)
+        return result
+
+    @staticmethod
+    @user_context
+    def update_theme(data, **kwargs):
+        """Updates theme."""
+        user: UserContext = kwargs["user"]
+        theme = Themes.get_theme(user.tenant_key)
```
Comment (shuhaib-aot) [inline forms-flow-api/src/formsflow_api/services/theme.py:37]: if Theme is not found what will be the response @AbijithS-aot @auslin-aot need to raise an exception right?

File: forms-flow-api/src/formsflow_api/resources/theme.py [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+        responses={
+            200: "OK:- Successful request.",
+            403: "FORBIDDEN:- Permission denied",
+        },
+        model=[theme_model],
+    )
+    def get():
+        """Get theme by tenant key. This is a public API."""
+        tenant_key = request.args.get("tenantKey", default=None)
+        response, status = (
```
Comment (shuhaib-aot) [inline forms-flow-api/src/formsflow_api/resources/theme.py:72]: if the theme is not found what will be a response @AbijithS-aot @auslin-aot

**Review by shuhaib-aot — COMMENTED** [review]
I'm uncertain whether we should automatically create a theme with default values for tenants when they are created. If we follow this approach, it would provide a basis for customization and allow us to offer a reset option in the future if needed.
@sumesh-aot  @AbijithS-aot  @auslin-aot

File: forms-flow-api/src/formsflow_api/services/theme.py (hunk 10 above)
Comment (auslin-aot) [reply forms-flow-api/src/formsflow_api/services/theme.py:37]: yes

**Comment by sumesh-aot** [comment]
> I'm uncertain whether we should automatically create a theme with default values for tenants when they are created. If we follow this approach, it would provide a basis for customization and allow us to offer a reset option in the future if needed. @sumesh-aot @AbijithS-aot @auslin-aot

@shuhaib-aot We can do that as part of tenant onboarding. We can keep these endpoints like this and leverage them from admin-api.

**Review by shuhaib-aot — APPROVED** [review]

**Pull request merged**